File: spike/route_preflight.py
"""Pure, privacy-bounded exact-host route preflight contract."""

from dataclasses import dataclass
import ipaddress
import json
import re


SCHEMA_VERSION = 1
MAX_PAYLOAD_BYTES = 2048
MAX_DEADLINE_MS = 8_000
MAX_CANDIDATE_ROUTES = 4

CANDIDATE_ROUTES = frozenset(("system", "app_doh", "local_strategy", "owned_geph"))
OUTCOMES = frozenset(
    (
        "navigation_pending",
        "regional_access_denied",
        "edge_access_denied",
        "challenge_or_auth",
        "usable",
        "terminal_error",
    )
)

REASON_ACCEPTED = "accepted"
REASON_BINDING_MISMATCH = "binding_mismatch"
REASON_CANDIDATE_NOT_ALLOWED = "candidate_not_allowed"
REASON_OBSERVATION_OUTSIDE_DEADLINE = "observation_outside_deadline"
REASON_EXPIRED = "expired"
REASON_REPLAY = "replay"

_CAPABILITY_RE = re.compile(r"^[0-9a-f]{32}$")
_HOST_LABEL_RE = re.compile(r"^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$")
_JOB_FIELDS = frozenset(
    (
        "schema_version",
        "capability",
        "host",
        "candidate_routes",
        "issued_at_unix_ms",
        "deadline_unix_ms",
    )
)
_RESULT_FIELDS = frozenset(
    (
        "schema_version",
        "capability",
        "host",
        "candidate_route",
        "outcome",
        "observed_at_unix_ms",
    )
)


class RoutePreflightError(ValueError):
    def __init__(self, code):
        super().__init__(code)
        self.code = code


@dataclass(frozen=True)
class RoutePreflightJobV1:
    capability: str
    host: str
    candidate_routes: tuple[str, ...]
    issued_at_unix_ms: int
    deadline_unix_ms: int
    schema_version: int = SCHEMA_VERSION


@dataclass(frozen=True)
class RoutePreflightResultV1:
    capability: str
    host: str
    candidate_route: str
    outcome: str
    observed_at_unix_ms: int
    schema_version: int = SCHEMA_VERSION


@dataclass(frozen=True)
class RoutePreflightDecision:
    accepted: bool
    reason: str
    host: str
    candidate_route: str
    outcome: str


def _object(payload, fields):
    if isinstance(payload, str):
        raw = payload.encode("utf-8")
    elif isinstance(payload, bytes):
        raw = payload
    else:
        raise RoutePreflightError("invalid_shape")
    if not raw:
        raise RoutePreflightError("empty_input")
    if len(raw) > MAX_PAYLOAD_BYTES:
        raise RoutePreflightError("payload_too_large")
    try:
        value = json.loads(raw)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise RoutePreflightError("invalid_json") from error
    if not isinstance(value, dict) or frozenset(value) != fields:
        raise RoutePreflightError("invalid_shape")
    if type(value["schema_version"]) is not int or value["schema_version"] != SCHEMA_VERSION:
        raise RoutePreflightError("unsupported_version")
    return value


def _capability(value):
    if not isinstance(value, str) or not _CAPABILITY_RE.fullmatch(value):
        raise RoutePreflightError("invalid_capability")
    return value


def _host(value):
    if not isinstance(value, str) or not value.isascii() or value != value.strip():
        raise RoutePreflightError("invalid_host")
    host = value.lower()
    if host.endswith("."):
        host = host[:-1]
    try:
        ipaddress.ip_address(host)
    except ValueError:
        pass
    else:
        raise RoutePreflightError("invalid_host")
    labels = host.split(".")
    if (
        not host
        or len(host) > 253
        or len(labels) < 2
        or any(not _HOST_LABEL_RE.fullmatch(label) for label in labels)
    ):
        raise RoutePreflightError("invalid_host")
    return host


def parse_route_preflight_job_v1(payload):
    value = _object(payload, _JOB_FIELDS)
    capability = _capability(value["capability"])
    host = _host(value["host"])
    routes = value["candidate_routes"]
    if (
        not isinstance(routes, list)
        or not routes
        or len(routes) > MAX_CANDIDATE_ROUTES
        or len(set(routes)) != len(routes)
        or any(type(route) is not str or route not in CANDIDATE_ROUTES for route in routes)
    ):
        raise RoutePreflightError("invalid_candidate_routes")
    issued_at = value["issued_at_unix_ms"]
    if type(issued_at) is not int or issued_at <= 0:
        raise RoutePreflightError("invalid_issued_at")
    deadline = value["deadline_unix_ms"]
    if (
        type(deadline) is not int
        or deadline <= issued_at
        or deadline - issued_at > MAX_DEADLINE_MS
    ):
        raise RoutePreflightError("invalid_deadline")
    return RoutePreflightJobV1(
        capability=capability,
        host=host,
        candidate_routes=tuple(routes),
        issued_at_unix_ms=issued_at,
        deadline_unix_ms=deadline,
    )


def parse_route_preflight_result_v1(payload):
    value = _object(payload, _RESULT_FIELDS)
    capability = _capability(value["capability"])
    host = _host(value["host"])
    route = value["candidate_route"]
    if type(route) is not str or route not in CANDIDATE_ROUTES:
        raise RoutePreflightError("invalid_candidate_route")
    outcome = value["outcome"]
    if type(outcome) is not str or outcome not in OUTCOMES:
        raise RoutePreflightError("invalid_outcome")
    observed_at = value["observed_at_unix_ms"]
    if type(observed_at) is not int or observed_at <= 0:
        raise RoutePreflightError("invalid_observed_at")
    return RoutePreflightResultV1(
        capability=capability,
        host=host,
        candidate_route=route,
        outcome=outcome,
        observed_at_unix_ms=observed_at,
    )


def validate_route_preflight_result_v1(
    job, result, *, now_unix_ms, capability_seen=False
):
    if capability_seen:
        reason = REASON_REPLAY
    elif (
        job.schema_version != result.schema_version
        or job.capability != result.capability
        or job.host != result.host
    ):
        reason = REASON_BINDING_MISMATCH
    elif result.candidate_route not in job.candidate_routes:
        reason = REASON_CANDIDATE_NOT_ALLOWED
    elif not job.issued_at_unix_ms <= result.observed_at_unix_ms <= job.deadline_unix_ms:
        reason = REASON_OBSERVATION_OUTSIDE_DEADLINE
    elif now_unix_ms > job.deadline_unix_ms:
        reason = REASON_EXPIRED
    else:
        reason = REASON_ACCEPTED
    return RoutePreflightDecision(
        accepted=reason == REASON_ACCEPTED,
        reason=reason,
        host=result.host,
        candidate_route=result.candidate_route,
        outcome=result.outcome,
    )

File: crates/slipstream-core/src/route_preflight.rs
//! Strict, privacy-bounded exact-host route preflight contract.
//!
//! A preflight result is evidence only. Route policy remains responsible for
//! comparing independent candidates and authorizing an exact-host transition.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use std::collections::HashSet;
use std::net::IpAddr;
use std::str::FromStr;

pub const ROUTE_PREFLIGHT_VERSION: u8 = 1;
pub const MAX_ROUTE_PREFLIGHT_BYTES: usize = 2048;
pub const MAX_ROUTE_PREFLIGHT_DEADLINE_MS: u64 = 8_000;
pub const MAX_ROUTE_PREFLIGHT_CANDIDATES: usize = 4;

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RoutePreflightCandidate {
    System,
    AppDoh,
    LocalStrategy,
    OwnedGeph,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RoutePreflightOutcome {
    NavigationPending,
    RegionalAccessDenied,
    EdgeAccessDenied,
    ChallengeOrAuth,
    Usable,
    TerminalError,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct RoutePreflightJobV1 {
    pub schema_version: u8,
    pub capability: String,
    pub host: String,
    pub candidate_routes: Vec<RoutePreflightCandidate>,
    pub issued_at_unix_ms: u64,
    pub deadline_unix_ms: u64,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct RoutePreflightResultV1 {
    pub schema_version: u8,
    pub capability: String,
    pub host: String,
    pub candidate_route: RoutePreflightCandidate,
    pub outcome: RoutePreflightOutcome,
    pub observed_at_unix_ms: u64,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RoutePreflightErrorCode {
    EmptyInput,
    PayloadTooLarge,
    InvalidJson,
    InvalidShape,
    UnsupportedVersion,
    InvalidCapability,
    InvalidHost,
    InvalidCandidateRoutes,
    InvalidCandidateRoute,
    InvalidOutcome,
    InvalidIssuedAt,
    InvalidDeadline,
    InvalidObservedAt,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RoutePreflightError {
    pub code: RoutePreflightErrorCode,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RoutePreflightDecisionReason {
    Accepted,
    BindingMismatch,
    CandidateNotAllowed,
    ObservationOutsideDeadline,
    Expired,
    Replay,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct RoutePreflightDecision {
    pub accepted: bool,
    pub reason: RoutePreflightDecisionReason,
    pub host: String,
    pub candidate_route: RoutePreflightCandidate,
    pub outcome: RoutePreflightOutcome,
}

fn error(code: RoutePreflightErrorCode) -> RoutePreflightError {
    RoutePreflightError { code }
}

fn parse_object(
    payload: &[u8],
    required_fields: &[&str],
) -> Result<Map<String, Value>, RoutePreflightError> {
    if payload.is_empty() {
        return Err(error(RoutePreflightErrorCode::EmptyInput));
    }
    if payload.len() > MAX_ROUTE_PREFLIGHT_BYTES {
        return Err(error(RoutePreflightErrorCode::PayloadTooLarge));
    }
    let value: Value =
        serde_json::from_slice(payload).map_err(|_| error(RoutePreflightErrorCode::InvalidJson))?;
    let object = value
        .as_object()
        .ok_or_else(|| error(RoutePreflightErrorCode::InvalidShape))?;
    if object.len() != required_fields.len()
        || !required_fields
            .iter()
            .all(|field| object.contains_key(*field))
    {
        return Err(error(RoutePreflightErrorCode::InvalidShape));
    }
    Ok(object.clone())
}

fn parse_version(object: &Map<String, Value>) -> Result<u8, RoutePreflightError> {
    let version = object["schema_version"]
        .as_u64()
        .and_then(|value| u8::try_from(value).ok())
        .ok_or_else(|| error(RoutePreflightErrorCode::UnsupportedVersion))?;
    if version != ROUTE_PREFLIGHT_VERSION {
        return Err(error(RoutePreflightErrorCode::UnsupportedVersion));
    }
    Ok(version)
}

fn parse_capability(object: &Map<String, Value>) -> Result<String, RoutePreflightError> {
    let capability = object["capability"]
        .as_str()
        .ok_or_else(|| error(RoutePreflightErrorCode::InvalidCapability))?;
    if capability.len() != 32
        || !capability
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
    {
        return Err(error(RoutePreflightErrorCode::InvalidCapability));
    }
    Ok(capability.to_owned())
}

fn normalize_hostname(value: &str) -> Option<String> {
    if value.trim() != value {
        return None;
    }
    let mut host = value.to_ascii_lowercase();
    if host.ends_with('.') {
        host.pop();
    }
    if host.is_empty() || host.len() > 253 || !host.contains('.') {
        return None;
    }
    if IpAddr::from_str(&host).is_ok() {
        return None;
    }
    let valid = host.split('.').all(|label| {
        !label.is_empty()
            && label.len() <= 63
            && !label.starts_with('-')
            && !label.ends_with('-')
            && label
                .bytes()
                .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || byte == b'-')
    });
    valid.then_some(host)
}

fn parse_host(object: &Map<String, Value>) -> Result<String, RoutePreflightError> {
    object["host"]
        .as_str()
        .and_then(normalize_hostname)
        .ok_or_else(|| error(RoutePreflightErrorCode::InvalidHost))
}

pub fn parse_route_preflight_job_v1(
    payload: &[u8],
) -> Result<RoutePreflightJobV1, RoutePreflightError> {
    const FIELDS: [&str; 6] = [
        "schema_version",
        "capability",
        "host",
        "candidate_routes",
        "issued_at_unix_ms",
        "deadline_unix_ms",
    ];
    let object = parse_object(payload, &FIELDS)?;
    let schema_version = parse_version(&object)?;
    let capability = parse_capability(&object)?;
    let host = parse_host(&object)?;
    let candidate_routes: Vec<RoutePreflightCandidate> =
        serde_json::from_value(object["candidate_routes"].clone())
            .map_err(|_| error(RoutePreflightErrorCode::InvalidCandidateRoutes))?;
    let unique: HashSet<_> = candidate_routes.iter().copied().collect();
    if candidate_routes.is_empty()
        || candidate_routes.len() > MAX_ROUTE_PREFLIGHT_CANDIDATES
        || unique.len() != candidate_routes.len()
    {
        return Err(error(RoutePreflightErrorCode::InvalidCandidateRoutes));
    }
    let issued_at_unix_ms = object["issued_at_unix_ms"]
        .as_u64()
        .filter(|value| *value > 0)
        .ok_or_else(|| error(RoutePreflightErrorCode::InvalidIssuedAt))?;
    let deadline_unix_ms = object["deadline_unix_ms"]
        .as_u64()
        .filter(|value| {
            *value > issued_at_unix_ms
                && value.saturating_sub(issued_at_unix_ms) <= MAX_ROUTE_PREFLIGHT_DEADLINE_MS
        })
        .ok_or_else(|| error(RoutePreflightErrorCode::InvalidDeadline))?;
    Ok(RoutePreflightJobV1 {
        schema_version,
        capability,
        host,
        candidate_routes,
        issued_at_unix_ms,
        deadline_unix_ms,
    })
}

pub fn parse_route_preflight_result_v1(
    payload: &[u8],
) -> Result<RoutePreflightResultV1, RoutePreflightError> {
    const FIELDS: [&str; 6] = [
        "schema_version",
        "capability",
        "host",
        "candidate_route",
        "outcome",
        "observed_at_unix_ms",
    ];
    let object = parse_object(payload, &FIELDS)?;
    let schema_version = parse_version(&object)?;
    let capability = parse_capability(&object)?;
    let host = parse_host(&object)?;
    let candidate_route = serde_json::from_value(object["candidate_route"].clone())
        .map_err(|_| error(RoutePreflightErrorCode::InvalidCandidateRoute))?;
    let outcome = serde_json::from_value(object["outcome"].clone())
        .map_err(|_| error(RoutePreflightErrorCode::InvalidOutcome))?;
    let observed_at_unix_ms = object["observed_at_unix_ms"]
        .as_u64()
        .filter(|value| *value > 0)
        .ok_or_else(|| error(RoutePreflightErrorCode::InvalidObservedAt))?;
    Ok(RoutePreflightResultV1 {
        schema_version,
        capability,
        host,
        candidate_route,
        outcome,
        observed_at_unix_ms,
    })
}

pub fn validate_route_preflight_result_v1(
    job: &RoutePreflightJobV1,
    result: &RoutePreflightResultV1,
    now_unix_ms: u64,
    capability_seen: bool,
) -> RoutePreflightDecision {
    let reason = if capability_seen {
        RoutePreflightDecisionReason::Replay
    } else if job.schema_version != result.schema_version
        || job.capability != result.capability
        || job.host != result.host
    {
        RoutePreflightDecisionReason::BindingMismatch
    } else if !job.candidate_routes.contains(&result.candidate_route) {
        RoutePreflightDecisionReason::CandidateNotAllowed
    } else if result.observed_at_unix_ms < job.issued_at_unix_ms
        || result.observed_at_unix_ms > job.deadline_unix_ms
    {
        RoutePreflightDecisionReason::ObservationOutsideDeadline
    } else if now_unix_ms > job.deadline_unix_ms {
        RoutePreflightDecisionReason::Expired
    } else {
        RoutePreflightDecisionReason::Accepted
    };
    RoutePreflightDecision {
        accepted: reason == RoutePreflightDecisionReason::Accepted,
        reason,
        host: result.host.clone(),
        candidate_route: result.candidate_route,
        outcome: result.outcome,
    }
}
